implementation/src/main/java/io/smallrye/mutiny/Context.java

```diff
@@ -0,0 +1,228 @@
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import io.smallrye.common.annotation.Experimental;
+
+/**
+ * A context allows sharing key / value entries along with a subscriber in a Mutiny pipeline, so all operators can
+ * share implicit data for a given subscription.
+ * <p>
+ * A context is provided by a {@link io.smallrye.mutiny.subscription.UniSubscriber} or {@link org.reactivestreams.Subscriber}
+ * that implements {@link io.smallrye.mutiny.subscription.ContextSupport}.
+ * <p>
+ * Context keys are represented as {@link String} while values can be from heterogeneous types.
+ * <p>
+ * {@link Context} instances are thread-safe.
+ * Internal storage is not allocated until the first entry is being added.
+ * <p>
+ * Contexts shall be primarily used to share transient data used for networked I/O processing such as correlation
+ * identifiers, tokens, etc.
+ * They should not be used as general-purpose data structures that are frequently updated and that hold large amounts of
+ * data.
+ *
+ * @see Uni#withContext(BiFunction)
+ * @see Uni#attachContext()
+ * @see io.smallrye.mutiny.groups.UniSubscribe
+ * @see Multi#withContext(BiFunction)
+ * @see Multi#attachContext()
+ * @see io.smallrye.mutiny.groups.MultiSubscribe
+ */
+@Experimental("Context support is a new experimental API introduced in Mutiny 1.3.0")
+public final class Context {
+
+    /**
+     * Creates a new empty context.
+     *
+     * @return the context
+     */
+    public static Context empty() {
+        return new Context();
+    }
+
+    /**
+     * Creates a context from key / value pairs.
+     *
+     * @param entries a sequence of key / value pairs, as in {@code key1, value1, key2, value2}
+     * @return the new context
+     * @throws IllegalArgumentException when {@code entries} is not balanced
+     * @throws NullPointerException when {@code entries} is {@code null}
+     */
+    public static Context of(Object... entries) {
+        requireNonNull(entries, "The entries array cannot be null");
+        if (entries.length % 2 != 0) {
+            throw new IllegalArgumentException("Arguments must be balanced to form (key, value) pairs");
+        }
+        HashMap<String, Object> map = new HashMap<>();
+        for (int i = 0; i < entries.length; i = i + 2) {
+            String key = nonNull(entries[i], "key").toString();
+            Object value = nonNull(entries[i + 1], "value");
+            map.put(key, value);
+        }
+        return new Context(map);
+    }
+
+    /**
+     * Creates a context by copying the entries from a {@link Map}.
+     *
+     * @param entries the map, cannot be {@code null}
+     * @return the new context
+     * @throws NullPointerException when {@code entries} is null
+     */
+    public static Context from(Map<String, ?> entries) {
+        return new Context(requireNonNull(entries, "The entries map cannot be null"));
+    }
+
+    private volatile ConcurrentHashMap<String, Object> entries;
```

Comment:
do you really need them to be both `volatile` and concurrent?

Reply:
In theory we could have 2 concurrent `put` that'd cause the initialisation

Reply:
Yes I understand the need with the current design; but I'm curious about 

1. why the whole Context needs to be concurrent-safe (even for initialization and writes)
2. why you opted for lazy initialization of the underlying map (I do love the attention towards allocations but I wonder how likely this is to not be allocated - and if it's very low, getting rid of this could simplify such code)
3. did you consider immutable contexts? Mutations could be modelled as a new view over an existing context (admittedly this would be less flexible and only suitable for relatively small collections but it could be fine - so it's a deeper question about intended use)

Reply:
1. Since you have 1 `Context` instance for a whole subscription lifespan, there are cases where events and subscriptions get offset to an executor. Or you merge pipelines (e.g., 2 concurrent HTTP requests), etc. There is 0 guarantee we won't have concurrent access for such kind of object, hence IMHO it's better to make it concurrent-proof from the start.
2. The code is more elegant / safer with non-`null` `Context` objects around, for instance some subscribers provide an empty context rather than flying a `null` around. Deferring the allocation is indeed all about reducing the footprint: subscription that use a context will allocate, those who don't will not pay the full price.
3. My first iterations had been around read / write views, but I found the developer experience to be worse than it should be. First off immutable contexts are nice on the paper, but you need to remember that the binding methods are called at subscription time, which is counter-intuitive (events from from publishers to subscribers, but here contexts flow from subscribers to publishers). I found out that have `withContext` (and `attachContext` as sugar) to bind was simple and flexible (no need to patch tons of methods) and has clear semantics. Then I found out that there was little value in read/write views, if you are after manipulating a context across a whole pipeline then making it a shared, thread-safe object was more intuitive.

(hope it makes sense, lots of points in a single post 😄 )

Reply:
Many thanks for all thourough explanations! Makes sense.